Rubberduck.Core/UI/CodeExplorer/CodeExplorerAddComponentService.cs

```diff
@@ -0,0 +1,122 @@
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using Rubberduck.VBEditor.Utility;
+
+namespace Rubberduck.UI.CodeExplorer
+{
+    public class CodeExplorerAddComponentService : ICodeExplorerAddComponentService
+    {            
+        private readonly IParseManager _parseManager;
+        private readonly IAddComponentService _addComponentService;
+        private readonly IVBE _vbe;
+
+        public CodeExplorerAddComponentService(IParseManager parseManager, IAddComponentService addComponentService, IVBE vbe)
+        {
+            _parseManager = parseManager;
+            _addComponentService = addComponentService;
+            _vbe = vbe;
+        }
+
+        public void AddComponent(CodeExplorerItemViewModel node, ComponentType componentType, string code = null)
+        {
+            var projectId = ProjectId(node);
+            if (projectId == null)
+            {
+                return;
+            }
+
+            var prefixInModule = FolderAnnotation(node);
+
+            _parseManager.OnSuspendParser(
+                this,
+                Enum.GetValues(typeof(ParserState)).Cast<ParserState>(),
+                () => _addComponentService.AddComponent(projectId, componentType, code, prefixInModule));
+        }
+
+        public void AddComponentWithAttributes(CodeExplorerItemViewModel node, ComponentType componentType, string code, string additionalPrefixInModule = null)
+        {
+            var projectId = ProjectId(node);
+            if (projectId == null)
+            {
+                return;
+            }
+
+            var folderAnnotation = FolderAnnotation(node);
+            var optionCompare = OptionCompareStatement();
+
+            var basePrefixInModule = optionCompare != null
+                ? $"{folderAnnotation}{Environment.NewLine}{optionCompare}"
+                : folderAnnotation;
+
+            var prefixInModule = additionalPrefixInModule != null
+                ? $"{basePrefixInModule}{Environment.NewLine}{additionalPrefixInModule}"
+                : basePrefixInModule;
+
+            _parseManager.OnSuspendParser(
+                this,
+                Enum.GetValues(typeof(ParserState)).Cast<ParserState>(),
+                () => _addComponentService.AddComponentWithAttributes(projectId, componentType, code, prefixInModule));
+        }
+
+        private string ProjectId(CodeExplorerItemViewModel node)
+        {
+            return node?.Declaration.ProjectId;
+        }
+
+        private string FolderAnnotation(CodeExplorerItemViewModel node)
+        {
+            return (node is CodeExplorerCustomFolderViewModel folder) 
+                ? folder.FolderAttribute 
+                : $"'@Folder(\"{Folder(node)}\")";
+        }
+
+        private string Folder(CodeExplorerItemViewModel node)
+        {
+            var declaration = node?.Declaration;
+            if (declaration == null)
+            {
+                return ActiveProjectName();
```

Comment:
Shouldn't this be `ActiveProjectFolder`?

Reply:
Yes, it should.